### PR TITLE
Logs - Implementing the ability to Download Logs

### PR DESCRIPTION
**Pull request by hwong0305**
#288 

- Updated code to have pm2 output logs into a specific location and filename for each app.
- implementing API to get error and output logs from the server
- Added test for API

Logs Screenshot

### CODE REVIEW

File: src/api/logs.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import { getMappingByDomain } from '../lib/data'
+
+const logsRouter = express.Router()
+const { isProduction } = environment
+
+logsRouter.get('/err/:domain', (req, res) => {
+  const { domain } = req.params
+  const { fullDomain } = getMappingByDomain(domain)
+
+  if (!isProduction()) return res.send('OK')
```
Comment (joshgreenwell) [inline src/api/logs.ts]: Why not get the logs in dev? I don't have context but, a comment in the code might be nice for future devs.

**Comment by joshgreenwell** [comment]
Can you add a screenshot to the PR description? I'm curious about what this feature looks like on the UI.

File: src/api/logs.ts (hunk 1 above)
Comment (hwong0305) [reply src/api/logs.ts]: I updated the codebase to allow for access of logs in dev.  From the current codebase with c0d3.com, the proxy server will not run when running in dev. The user will have to navigate to the port on the server to access their apps.

**Comment by hwong0305** [comment]
Updated PR description to include screenshot of the feature

File: src/helpers/environment.ts [hunk 2]
```diff
@@ -5,5 +5,6 @@ export default {
   HOME: process.env.HOME || process.env.ENV || null,
   WORKPATH: process.env.WORKPATH || '/home/myproxy',
   isProduction: (): boolean =>
-    (process.env.NODE_ENV || process.env.ENV) === 'production'
+    (process.env.NODE_ENV || process.env.ENV) === 'production',
+  isTest: (): boolean => (process.env.NODE_ENV || process.env.ENV) === 'test'
```
Comment (songz) [inline src/helpers/environment.ts]: this means that during development, we have to run this with test now? I don't think devs are ready for that yet. Could we use production flag?

File: src/public/client.ts [hunk 3]
```diff
@@ -41,15 +41,18 @@ class MappingItem {
     const mappingElement = document.createElement('li')
     let iconClass
     let iconColor
+    let logClass
```
Comment (songz) [inline src/public/client.ts]: `let logClass = 'fa fa-file-text-o ml-1 mt-1'`

Comment (songz) [reply src/public/client.ts]: why does this even have to be a variable? and putting this inside the `status` check block is kinda strange, what does status have to do with logs?

File: src/tests/helpers/logAdapter.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  const options: Options = {
+    method,
+    headers: reqHeaders
+  }
+  if (body) {
+    options.body = JSON.stringify(body)
+  }
+  if (target === 'mappings') {
+    return fetch(`${apiURL}/api/mappings${path}`, options)
+  }
```
Comment (songz) [inline src/tests/helpers/logAdapter.ts]: remove lines 25-27

File: src/public/client.ts (hunk 3 above)
Comment (hwong0305) [reply src/public/client.ts]: This to disable UI for getting logs when pm2 doesn't return a status. (i.e for c0d3.com's application)

File: src/api/logs.ts (hunk 1 above)
Comment (songz) [reply src/api/logs.ts]: @joshgreenwell `logs` are directly tied to `root pm2`, so one should be running in production flag when testing these to simulate production as much as possible.

At this point I don't see any need other than prod vs non-prod.

Usually `test` vs `stage` vs `sandbox` environment comes into importance when you start dealing with databases and billing.

**Pull request merged**